File: leetcode/201-300/202-happy-number/happy_number.py
"""LeetCode 202 — Happy Number (Python mirror / oracle).

Floyd's tortoise-and-hare over the sum-of-squared-digits map; happy iff the fast
pointer reaches 1. Mirrors the Kāra version.
"""


def sq_digit_sum(n):
    total = 0
    x = n
    while x > 0:
        d = x % 10
        total += d * d
        x //= 10
    return total


def is_happy(n):
    slow = n
    fast = sq_digit_sum(n)
    while fast != 1 and slow != fast:
        slow = sq_digit_sum(slow)
        fast = sq_digit_sum(sq_digit_sum(fast))
    return fast == 1


def report(n):
    print("true" if is_happy(n) else "false")


def main():
    report(19)
    report(2)
    report(1)
    report(7)
    report(4)
    report(100)
    report(1111111)


main()

File: leetcode/201-300/202-happy-number/bench/happy_number.rs
fn sq_digit_sum(n: i64) -> i64 {
    let mut total = 0i64;
    let mut x = n;
    while x > 0 {
        let d = x % 10;
        total += d * d;
        x /= 10;
    }
    total
}

fn is_happy(n: i64) -> bool {
    let mut slow = n;
    let mut fast = sq_digit_sum(n);
    while fast != 1 && slow != fast {
        slow = sq_digit_sum(slow);
        fast = sq_digit_sum(sq_digit_sum(fast));
    }
    fast == 1
}

fn main() {
    let limit: i64 = 4000000;
    let mut sink: i64 = 0;
    for i in 1..limit {
        if is_happy(i) {
            sink += 1;
        }
    }
    println!("{}", sink);
}
